### ISSUE TITLE
[Windows Forms] Controls support

### ISSUE THREAD

**Issue by kant2002** [dotnet/runtimelab#1100]
Here the summary of controls working in Windows Forms in my limited testing

# Common Controls
- [X] Button
- [X] CheckBox
- [X] CheckedListBox
- [x] ComboBox
- [x] DateTimePicker
- [X] Label
- [X] LinkLabel
- [X] ListBox
- [X] ListView
- [X] MaskedTextBox
- [x] MonthCalendar Crash on startup
- [x] NotifyIcon
- [X] NumericUpDown
- [X] PictureBox
- [X] ProgressBar
- [x] PropertyGrid
- [X] RadioButton
- [x] RichTextBox. 
- [X] TextBox
- [X] ToolTip
- [X] TrackBar
- [X] TreeView
- [ ] WebBrowser Only events are missing I believe.

# Containers
- [X] FlowLayoutPanel
- [X] GroupBox
- [X] Panel
- [X] SplitContainer 
- [X] TabControl
- [X] TableLayoutPanel

# Menu & Toolstrip
- [X] ContextMenuStrip
- [X] MenuStrip
- [X] StatusStrip
- [X] ToolStrip
- [X] ToolStripContainer
- [x] ToolStripSplitButton

# Dialogs
- [x] OpenFileDialog. on .NET 6 only. Nothing can be done for .NET 5 apps
- [x] SaveFileDialog. on .NET 6 only. Nothing can be done for .NET 5 apps
- [ ] FolderBrowserDialog. Need support for Marshal.FinalReleaseComObject in NativeAOT or removal of this pattern in WinForms repo.
- [x] FontDialog.
- [x] ColorDialog.

# DataGridView
- [X] DataGridView.

<s>All crashes so far is from loading images from stream and require support for `Interop.Ole32.IStream` from `System.Drawing.Common`</s>

This is resurrection of https://github.com/dotnet/corert/issues/7995 and all tests was done with `WinFormsComWrappers` from https://github.com/kant2002/WinFormsComInterop/

<s>To make combobox works require implementation of `Accessibility.IAccessible`. </s>

**Comment by kant2002** [comment]
Combobox is working now, but with broken accessibility.

**Comment by kant2002** [comment]
Combobox is working now.

**Comment by weltkante** [comment]
> Most likely exception and hopefully not due to missing IDispatch support.

I'm pretty sure IDispatch is needed for WebBrowser scripting and eventing support (and besides WebBrowser, also for some Clipboard / Drag'n'drop scenarios; also for ActiveX hosting scenarios, both ways, consuming and providing controls). We had several recent issues due to changes in how IDispatch works in WinForms in .NET Core (its no longer provided by base classes, users have to request it explicitly now). So its clearly needed and people are using it.

You may be able to run a limited form of WinForms this way, but definitely cannot get it fully working without proper COM support. It may be possible to roll custom COM support into WinForms (for e.g. IDispatch and whatever else is missing) but its probably still quite some work.

**Comment by JensNordenbro** [comment]
The work that is related to this item is for NativeAOT as I understand it; not the AOT (PublishReadyToRun) of NET6 today and related to COM issues. 
Today it is not possible to trim a publish of a net6 winforms/wpf setup. This also seems to be dependent upon COM issues. 

Will this work improve on NET6 trimming capability in current days AOT or does this only apply to NativeAOT?

**Comment by kant2002** [comment]
This work does not affect trimming in any way. At least I think that's separate issue.

> This also seems to be dependent upon COM issues.

If not blocked by tooling, you may have similar experience with `PublishReadyToRun` and trimming, since trimming right now disable COM support. You have two paths from here, enable com support using `BuiltInComInteropSupport`, unlikely improve size of your app. Second path, is to add [custom ComWrappers instance](https://github.com/kant2002/WinFormsComInterop/). Obviously you can write ComWrappers yourself and not rely on my library.

> Will this work improve on NET6 trimming capability in current days AOT or does this only apply to NativeAOT?

Any further work on WinForms would require improve trimming of WinForms itself, and would happens in WinForms repo [itself](https://github.com/dotnet/winforms/issues/4649). So I would say, any NativeAOT work and ReadyToRun work would go side-by-side and improve both at the same time. At least I see it this way now.

Other option which may help you, is to use custom build of WinForms where you remove all not used code yourself. With WinForms that's extremely simple to do.

WPF is out of discussion for now, due to [lack of resources](https://github.com/dotnet/wpf/issues/3811) (IMO) in supporting team.

**Comment by JensNordenbro** [comment]
Just to clarify; I assume the gap in webbrowsing component is related to nuget Microsoft.Web.WebView2 ?

**Comment by kant2002** [comment]
No, this is regular `System.Windows.Forms.WebControls`. It's hard to make that works and not break existing code. Somewhat that control is working, but I see gaps in the supporting features. New WebView2 support is here https://github.com/MicrosoftEdge/WebView2Feedback/issues/1490

**Comment by jkotas** [comment]
WinForms support for Native AOT is tracked by https://github.com/dotnet/winforms/issues/4649
